torch/fx/graph_module.py

```diff
@@ -277,6 +266,165 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+                other attribute)
+
+        """
+        *prefix, field = target.split('.')
+        mod: torch.nn.Module = self
+
+        for item in prefix:
+
+            submod = getattr(self, item, None)
+
+            if submod is None:
+                submod = torch.nn.Module()
+                setattr(mod, item, submod)
+
+            if not isinstance(submod, torch.nn.Module):
+                return False
+
+            mod = submod
+
+        setattr(self, field, m)
+        return True
+
+    def delete_submodule(self, target: str) -> bool:
+        """
+        Deletes the given submodule from ``self``.
+
+        The module will not be deleted if ``target`` is not a valid
+        target.
+
+        Args:
+            target: The fully-qualified string name of the new submodule
+                (See example in ``has_submodule`` for how to denote a
+                fully-qualified string.)
+
+        Returns:
+            bool: Whether or not the target string referenced a
+                submodule we want to delete. A return value of ``False``
+                means that the ``target`` was not a valid reference to
+                a submodule.
+        """
+        def find_and_delete(o: Any, target: str) -> bool:
+            atoms = target.split(".", 1)
+            prefix, path = atoms[0], atoms[1] if len(atoms) > 1 else None
+
+            if not hasattr(o, prefix):
+                return False
+
+            # If the submodule we're looking for is on this layer
+            if not path:
+                if not isinstance(getattr(o, prefix), torch.nn.Module):
+                    return False
+
+                delattr(o, prefix)
+                return True
+
+            return find_and_delete(getattr(o, prefix), path)
+
+        return find_and_delete(self, target)
+
+    def delete_all_unused_submodules(self) -> None:
+        """
+        Deletes all unused submodules from ``self``.
+
+        A Module is considered "used" if any one of the following is
+        true:
+        1. It has children that are used
+        2. Its forward is called directly via a ``call_module`` node
+        3. It has a non-Module attribute that is used from a 
+           ``get_attr`` node
+
+        This method can be called to clean up a GraphModule without
+        manually calling ``delete_submodule`` on each unused submodule.
+        """
+        # Collect all the call_module and get_attr targets as well as 
+        # the names of their intermediary modules. For example, if we 
+        # have the target `foo.bar.baz`, we'll add `foo`, `foo.bar`,
+        # and `foo.bar.baz` to the list
+        used: List[str] = [name for node in self.graph.nodes 
+                           if node.op == "call_module" or node.op == "get_attr"
+                           for name in itertools.accumulate(node.target.split("."),
+                                                            lambda x, y: x + "." + y if y else x)]
```

Comment:
This is pretty gnarly to read. Can we split this up into nested loops/if statements and have a clear comment about what the accumulate bit is doing (I believe it's registering all prefixes for every qualified name but it takes a while to parse that out